File: apps/api/internal/model/signal.go
package model

import "time"

// SignalType represents the type of trading signal.
type SignalType string

const (
	SignalBullish SignalType = "bullish"
	SignalBearish SignalType = "bearish"
	SignalNeutral SignalType = "neutral"
)

// Signal represents a computed trading signal.
type Signal struct {
	Name        string     `json:"name"`
	Type        SignalType `json:"type"`
	Score       float64    `json:"score"`
	Description string     `json:"description"`
}

// SignalResult contains all computed signals for a stock.
type SignalResult struct {
	Ticker          string    `json:"ticker"`
	Signals         []Signal  `json:"signals"`
	PiotroskiScore  int       `json:"piotroskiScore"`
	OverallSentiment SignalType `json:"overallSentiment"`
	ComputedAt      time.Time `json:"computedAt"`
}

File: packages/shared/src/types/signal.ts
/**
 * Signal types indicating market sentiment.
 */
export type SignalType = 'bullish' | 'bearish' | 'neutral';

/**
 * Individual trading signal.
 */
export interface Signal {
  /** Signal identifier/name */
  name: string;
  /** Overall sentiment */
  type: SignalType;
  /** Numeric score (typically 0-1 or 0-100) */
  score: number;
  /** Human-readable explanation */
  description: string;
}

/**
 * Piotroski F-Score breakdown.
 * Each component is 0 or 1, total score 0-9.
 */
export interface PiotroskiScore {
  /** Total score (0-9) */
  total: number;
  /** Profitability signals (0-4) */
  profitability: {
    positiveROA: boolean;
    positiveOperatingCashFlow: boolean;
    roaImprovement: boolean;
    cashFlowGreaterThanNetIncome: boolean;
  };
  /** Leverage/liquidity signals (0-3) */
  leverage: {
    decreasedLeverage: boolean;
    increasedCurrentRatio: boolean;
    noNewShares: boolean;
  };
  /** Operating efficiency signals (0-2) */
  efficiency: {
    improvedGrossMargin: boolean;
    improvedAssetTurnover: boolean;
  };
}

/**
 * Complete signal analysis result.
 */
export interface SignalResult {
  ticker: string;
  signals: Signal[];
  piotroskiScore: PiotroskiScore;
  overallSentiment: SignalType;
  computedAt: string;
}
